scan.go

```diff
@@ -48,308 +44,27 @@ var (
-	if err != nil {
-		txn.Rollback()
-		return 0, err
-	}
-	qry, err := txn.Prepare(`SELECT 1 FROM scan WHERE ip=? AND port=? AND proto=?`)
-	if err != nil {
-		txn.Rollback()
-		return 0, err
-	}
-	update, err := txn.Prepare(`UPDATE scan SET lastseen=? WHERE ip=? AND port=? AND proto=?`)
-	if err != nil {
-		txn.Rollback()
-		return 0, err
-	}
-
-	var count int64
-
-	for _, r := range results {
-		// Although it's an array, only one port is in each
-		port := r.Ports[0]
-
-		// Skip results which are (usually) banner-only
-		// While it would be nice to store banners, we need to restructure a
-		// bit to accommodate this and it just inserts duplicate data for now
-		if port.Status == "" || port.Service.Name != "" {
-			continue
-		}
-
-		// Search for the IP/port/proto combo
-		// If it exists, update `lastseen`, else insert a new record
-
-		// Because we have to scan into something
-		var x int
-		err := qry.QueryRow(r.IP, port.Port, port.Proto).Scan(&x)
-		switch {
-		case err == sql.ErrNoRows:
-			_, err = insert.Exec(r.IP, port.Port, port.Proto, now, now)
-			if err != nil {
-				txn.Rollback()
-				return 0, err
-			}
-			count++
-			continue
-		case err != nil:
-			txn.Rollback()
-			return 0, err
-		}
-
-		_, err = update.Exec(now, r.IP, port.Port, port.Proto)
-		if err != nil {
-			txn.Rollback()
-			return 0, err
-		}
-
-		count++
-	}
-
-	txn.Commit()
-	return count, nil
-}
-
-func loadSubmission(filter sqlFilter) (submission, error) {
-	var host string
-	var job sql.NullInt64
-	var subTime NullTime
-
-	qry := fmt.Sprintf(`SELECT host, job_id, submission_time FROM submission %s ORDER BY rowid DESC LIMIT 1`, filter)
-	err := db.QueryRow(qry, filter.Values...).Scan(&host, &job, &subTime)
-	if err != nil && err != sql.ErrNoRows {
-		log.Println("loadSubmission: error scanning table:", err)
-		return submission{}, err
-	}
-
-	return submission{Host: host, Job: job.Int64, Time: scanTime(subTime.Time.UTC())}, nil
-}
-
-func saveSubmission(host string, job *int64, now time.Time) error {
-	txn, err := db.Begin()
-	if err != nil {
-		return err
-	}
-
-	qry := `INSERT INTO submission (host, job_id, submission_time) VALUES (?, ?, ?)`
-	_, err = txn.Exec(qry, host, toNullInt64(job), now)
-	if err != nil {
-		txn.Rollback()
-		return err
-	}
-
-	err = txn.Commit()
-	if err != nil {
-		return err
-	}
-
-	if job != nil {
-		gaugeJobSubmission.Set(float64(now.Unix()))
-	} else {
-		gaugeSubmission.Set(float64(now.Unix()))
-	}
-
-	return nil
-}
-
-func loadTraceroutes() (map[string]struct{}, error) {
-	ips := make(map[string]struct{})
-
-	rows, err := db.Query(`SELECT dest FROM traceroute`)
-	if err != nil {
-		return nil, err
-	}
-	defer rows.Close()
-
-	var ip string
-	for rows.Next() {
-		err := rows.Scan(&ip)
-		if err != nil {
-			return nil, err
-		}
-		if _, ok := ips[ip]; !ok {
-			ips[ip] = struct{}{}
-		}
-	}
-
-	return ips, nil
+type storage interface {
+	LoadData(filter sqlite.SQLFilter) ([]scan.IPInfo, error)
```

Comment:
This (and other lines) need some fixing as it's directly using the `sqlite.SQLFilter` type in the interface, which isn't right, but I'll figure that out later.